File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/Syms/TypeAttributePersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.Syms
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.Syms.Models;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.Syms.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Linq;
    using Newtonsoft.Json.Serialization;
    using System;
    using System.Collections.Generic;
    using System.Linq;

    class TypeAttributePersistence
    {
        private static readonly string Tag = nameof(TypeAttributePersistence);

        public static CdmTypeAttributeDefinition FromData(CdmCorpusContext ctx, DataColumn obj, string entityName = null)
        {
            var typeAttribute = ctx.Corpus.MakeObject<CdmTypeAttributeDefinition>(CdmObjectType.TypeAttributeDef, obj.Name);
            var properties = obj.OriginDataTypeName.Properties;

            typeAttribute.DataFormat = Utils.SymsDataTypeToCDMDataFormat(obj.OriginDataTypeName);

            if (obj.OriginDataTypeName.Length > 0 && typeAttribute.DataFormat == CdmDataFormat.String)
            {
                typeAttribute.DataFormat = CdmDataFormat.Guid;
            }

            if (obj.OriginDataTypeName.Scale != 0 || obj.OriginDataTypeName.Precision != 0)
            {
                var numericTraits = ctx.Corpus.MakeRef<CdmTraitReference>(CdmObjectType.TraitRef, "is.dataFormat.numeric.shaped", true);

                var scaleTraitsArg = ctx.Corpus.MakeObject<CdmArgumentDefinition>(CdmObjectType.ArgumentDef, "scale");
                scaleTraitsArg.Value = obj.OriginDataTypeName.Scale;
                numericTraits.Arguments.Add(scaleTraitsArg);

                var precisionTraitsArg = ctx.Corpus.MakeObject<CdmArgumentDefinition>(CdmObjectType.ArgumentDef, "scale");
                precisionTraitsArg.Value = obj.OriginDataTypeName.Scale;
                numericTraits.Arguments.Add(precisionTraitsArg);
            }

            if (properties != null)
            {
                if (properties.ContainsKey("cdm:purpose"))
                {
                    typeAttribute.Purpose = properties["cdm:purpose"].ToObject<CdmPurposeReference>();
                }
                if (properties.ContainsKey("cdm:dataType"))
                {
                    typeAttribute.DataType = properties["cdm:dataType"].ToObject<CdmDataTypeReference> ();
                }
                if (properties.ContainsKey("cdm:cardinality"))
                {
                    string minCardinality = null;
                    if (properties["cdm:minimum"] != null)
                        minCardinality = properties["cdm:minimum"].ToObject<string>();

                    string maxCardinality = null;
                    if (properties["cdm:maximum"] != null)
                        maxCardinality = properties["cdm:maximum"].ToObject<string>();

                    if (string.IsNullOrWhiteSpace(minCardinality) || string.IsNullOrWhiteSpace(maxCardinality))
                    {
                        Logger.Error((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.ErrPersistCardinalityPropMissing);
                    }

                    if (!CardinalitySettings.IsMinimumValid(minCardinality))
                    {
                        Logger.Error((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.ErrValdnInvalidMinCardinality, minCardinality);
                    }

                    if (!CardinalitySettings.IsMaximumValid(maxCardinality))
                    {
                        Logger.Error((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.ErrValdnInvalidMaxCardinality, maxCardinality);
                    }

                    if (!string.IsNullOrWhiteSpace(minCardinality) && !string.IsNullOrWhiteSpace(maxCardinality) &&
                        CardinalitySettings.IsMinimumValid(minCardinality) && CardinalitySettings.IsMinimumValid(maxCardinality))
                    {
                        typeAttribute.Cardinality = new CardinalitySettings(typeAttribute)
                        {
                            Minimum = minCardinality,
                            Maximum = maxCardinality
                        };
                    }
                }
                if (properties.ContainsKey("cdm:traits"))
                {
                    Utils.AddListToCdmCollection(typeAttribute.AppliedTraits, Utils.CreateTraitReferenceList(ctx, properties["cdm:traits"]));
                }
                if (properties.ContainsKey("cdm:isPrimaryKey") &&
                    properties["cdm:isPrimaryKey"].ToObject<bool>())
                {
                    TraitToPropertyMap t2pMap = new TraitToPropertyMap(typeAttribute);
                    t2pMap.UpdatePropertyValue("isPrimaryKey", entityName + "/(resolvedAttributes)/" + typeAttribute.Name);
                }
                if (properties.ContainsKey("cdm:isReadOnly"))
                {
                    typeAttribute.IsReadOnly = properties["cdm:isReadOnly"].ToObject<bool>();
                }
                if (properties.ContainsKey("cdm:sourceName"))
                {
                    typeAttribute.SourceName = properties["cdm:sourceName"].ToObject<string>();
                }
                if (properties.ContainsKey("cdm:sourceOrdering"))
                {
                    typeAttribute.SourceOrdering = properties["cdm:sourceOrdering"].ToObject<int>();
                }
                if (properties.ContainsKey("cdm:valueConstrainedToList"))
                {
                    typeAttribute.ValueConstrainedToList = properties["cdm:valueConstrainedToList"].ToObject<bool>();
                }
                if (properties.ContainsKey("cdm:maximumLength"))
                {
                    typeAttribute.MaximumLength = properties["cdm:maximumLength"].ToObject<int>();
                }
                if (properties.ContainsKey("cdm:maximumValue"))
                {
                    typeAttribute.MaximumValue = properties["cdm:maximumValue"].ToObject<string>();
                }
                if (properties.ContainsKey("cdm:minimumValue"))
                {
                    typeAttribute.MinimumValue = properties["cdm:minimumValue"].ToObject<string>();
                }
                if (properties.ContainsKey("cdm:defaultValue"))
                {
                    typeAttribute.DefaultValue = properties["cdm:defaultValue"];
                }
            }
            
            return typeAttribute;
        }

        public static DataColumn ToData(CdmTypeAttributeDefinition instance, CdmCorpusContext ctx, ResolveOptions resOpt, CopyOptions options)
        {
            var properties = CreateProperties(instance, resOpt, options);
            
            var originDataTypeName = new TypeInfo
            {
                Properties = properties,
                IsComplexType = false,
                IsNullable = instance.GetProperty("isNullable"),
                TypeFamily = "cdm"
            };

            var t2pm = new TraitToPropertyMap(instance); ;
            var numericTraits = t2pm.FetchTraitReference("is.dataFormat.numeric.shaped");
            if (numericTraits != null)
            {
                foreach (var numericTraitsArg in numericTraits.Arguments)
                {
                    if (numericTraitsArg.Name == "precision")
                    {
                        originDataTypeName.Precision = (int)numericTraitsArg.Value;
                    }
                    if (numericTraitsArg.Name == "scale")
                    {
                        originDataTypeName.Scale = (int)numericTraitsArg.Value;
                    }
                }
            }

            var dataFormat = instance.GetProperty("dataFormat");
            originDataTypeName = Utils.CDMDataFormatToSymsDataType(dataFormat, originDataTypeName);
            if (originDataTypeName.TypeName == null)
            {
                Logger.Error(ctx, Tag, nameof(ToData), instance.AtCorpusPath, CdmLogCode.ErrPersistSymsUnknownDataFormat, instance.DisplayName);
                return null;
            }

            return new DataColumn
            {
                OriginDataTypeName = originDataTypeName,
                Name = instance.Name
            };
        }

        private static Dictionary<string, JToken> CreateProperties(CdmTypeAttributeDefinition instance, ResolveOptions resOpt, CopyOptions options)
        {
            var properties = new Dictionary<string, JToken>();

            var displayName = instance.GetProperty("displayName");
            var sourceName = instance.GetProperty("sourceName");
            var description = instance.GetProperty("description");
            var isReadOnly = instance.GetProperty("isReadOnly");
            var maximumLength = instance.GetProperty("maximumLength");
            var maximumValue = instance.GetProperty("maximumValue");
            var minimumValue = instance.GetProperty("minimumValue");
            var sourceOrdering = instance.GetProperty("sourceOrdering");
            var valueConstrainedToList = instance.GetProperty("valueConstrainedToList");
            var isPrimaryKey = instance.GetProperty("isPrimaryKey");
            var defValue = instance.GetProperty("defaultValue");

            if (displayName != null)
            {
                properties["cdm:displayName"] = JToken.FromObject(displayName);
            }

            if (instance.Explanation != null)
            {
                properties["cdm:explanation"] = JToken.FromObject(instance.Explanation);
            }

            if (sourceName != null)
            {
                properties["cdm:sourceName"] = JToken.FromObject(sourceName);
            }

            if (description != null)
            {
                properties["cdm:description"] = JToken.FromObject(description);
            }

            if (instance.AppliedTraits != null && instance.AppliedTraits.Count > 0)
            {
                properties["cdm:traits"] = JToken.FromObject(CopyDataUtils.ListCopyData(resOpt, instance.AppliedTraits?
                .Where(trait => trait is CdmTraitGroupReference || !(trait as CdmTraitReference).IsFromProperty), options),
                    new JsonSerializer { NullValueHandling = NullValueHandling.Ignore, ContractResolver = new CamelCasePropertyNamesContractResolver() });
            }

            if (isReadOnly != null)
            {
                properties["cdm:isReadOnly"] = JToken.FromObject(isReadOnly);
            }

            if (maximumLength != null)
            {
                properties["cdm:maximumLength"] = JToken.FromObject(maximumLength);
            }

            if (maximumValue != null)
            {
                properties["cdm:maximumValue"] = JToken.FromObject(maximumValue);
            }

            if (minimumValue != null)
            {
                properties["cdm:minimumValue"] = JToken.FromObject(minimumValue);
            }

            if (sourceOrdering != null && sourceOrdering != 0)
            {
                properties["cdm:sourceOrdering"] = JToken.FromObject(sourceOrdering);
            }

            if (valueConstrainedToList != null)
            {
                properties["cdm:valueConstrainedToList"] = JToken.FromObject(valueConstrainedToList);
            }

            if (isPrimaryKey != null)
            {
                properties["cdm:isPrimaryKey"] = JToken.FromObject(isPrimaryKey);
            }

            if (defValue != null)
            {
                properties["cdm:defaultValue"] = JToken.FromObject(defValue);
            }

            return properties;
        }
    }
}

File: objectModel/Python/cdm/persistence/syms/type_attribute_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Optional

from cdm.objectmodel import CdmCorpusContext, CdmTypeAttributeDefinition, CdmTraitGroupReference
from cdm.enums import CdmDataFormat, CdmObjectType
from cdm.utilities import logger, ResolveOptions, CopyOptions, TraitToPropertyMap, copy_data_utils
from cdm.enums import CdmLogCode
from cdm.persistence.syms.models import ColumnRelationshipInformation, DataColumn, DataSource, \
    DatabaseEntity, DatabaseProperties, FormatInfo, Namespace, PartitionInfo, PartitionInfoNamespace, \
    PartitionInfoProperties, RelationshipEntity, RelationshipProperties, SASEntityType, ScalarTypeInfo, \
    SchemaEntity, StorageDescriptor, TableEntity, TableNamespace, TablePartitioning, TableProperties, \
    TypeInfo

from . import utils
from .attribute_context_reference_persistence import AttributeContextReferencePersistence
from .attribute_resolution_guidance_persistence import AttributeResolutionGuidancePersistence
from .data_type_reference_persistence import DataTypeReferencePersistence
from .purpose_reference_persistence import PurposeReferencePersistence
from cdm.objectmodel.projections.cardinality_settings import CardinalitySettings
from .types import TypeAttribute
from cdm.persistence.syms.models import ColumnRelationshipInformation, DataColumn, DataSource, DatabaseEntity, \
    DatabaseProperties, FormatInfo, Namespace, PartitionInfo, PartitionInfoNamespace, PartitionInfoProperties, \
    RelationshipEntity, RelationshipProperties, ScalarTypeInfo, SchemaEntity, StorageDescriptor, TableEntity, \
    TableNamespace, TablePartitioning, TableProperties, TypeInfo

_TAG = 'TypeAttributePersistence'

class TypeAttributePersistence:
    @staticmethod
    def from_data(ctx: CdmCorpusContext, obj: DataColumn,
                  entity_name: Optional[str] = None) -> CdmTypeAttributeDefinition:
        type_attribute = ctx.corpus.make_object(CdmObjectType.TYPE_ATTRIBUTE_DEF, obj.name)
        properties = obj.origin_data_type_name.properties

        type_attribute.data_format = utils.syms_data_type_to_cdm_data_format(obj.origin_data_type_name)
        type_attribute.is_nullable = obj.origin_data_type_name.is_nullable

        if obj.origin_data_type_name.scale != 0 or obj.origin_data_type_name.precision != 0:
            numeric_traits = ctx.corpus.make_ref(CdmObjectType.TRAIT_REF, 'is.data_format.numeric.shaped', True)
        
            scale_traits_arg = ctx.corpus.make_ref(CdmObjectType.ARGUMENT_DEF, 'scale', False)
            scale_traits_arg.Value = obj.origin_data_type_name.scale
            numeric_traits.arguments.append(scale_traits_arg)
        
            precision_traits_arg = ctx.corpus.make_ref(CdmObjectType.ARGUMENT_DEF, 'scale', False)
            precision_traits_arg.value = obj.origin_data_type_name.scale
            numeric_traits.arguments.append(precision_traits_arg)

        if properties is not None:
            if 'cdm:purpose' in properties:
                type_attribute.purpose = properties['cdm:purpose']
            if 'cdm:dataType' in properties:
                type_attribute.data_type = properties['cdm:dataType']
            if 'cdm:traits' in properties:
                utils.add_list_to_cdm_collection(type_attribute.applied_traits, utils.create_trait_reference_array(ctx, properties['cdm:traits']))
            if 'cdm:is_primary_key' in properties and properties['cdm:is_primary_key']:
                t2pMap = TraitToPropertyMap(type_attribute)
                t2pMap._update_property_value('is_primary_key', entity_name + '/(resolvedAttributes)/' + type_attribute.name)
            if 'cdm:isReadOnly' in properties:
                type_attribute.is_read_only = properties['cdm:isReadOnly']
            if 'cdm:sourceName' in properties:
                type_attribute.source_name = properties['cdm:source_name']
            if 'cdm:sourceOrdering' in properties:
                type_attribute.source_ordering = properties['cdm:sourceOrdering']
            if 'cdm:valueConstrainedToList' in properties:
                type_attribute.value_constrained_to_list = properties['cdm:valueConstrainedToList']
            if 'cdm:maximumLength' in properties:
                type_attribute.maximum_length = properties['cdm:maximumLength']
            if 'cdm:maximumValue' in properties:
                type_attribute.maximum_value = properties['cdm:maximumValue']
            if 'cdm:minimumValue' in properties:
                type_attribute.minimum_value = properties['cdm:minimumValue']
            if 'cdm:defaultValue' in properties:
                type_attribute.default_value = properties['cdm:defaultValue']
            return type_attribute

    def to_data(instance: CdmTypeAttributeDefinition, ctx: CdmCorpusContext, res_opt: ResolveOptions, options: CopyOptions) -> TypeAttribute:
        properties = TypeAttributePersistence.create_properties(instance, res_opt, options)
        origin_data_type_name = TypeInfo(
        type_name = '',
        properties = properties,
        is_complex_type = False,
        is_nullable = instance._get_property('isNullable'),
        type_family = 'cdm')

        t2pm = TraitToPropertyMap(instance)
        numeric_traits = t2pm._fetch_trait_reference('is.data_format.numeric.shaped')
        if numeric_traits is not None:
            for numeric_traits_arg in numeric_traits.argument:
                if numeric_traits_arg.name == 'precision':
                    origin_data_type_name.precision = numeric_traits_arg.value
                if numeric_traits_arg.Name == 'scale':
                    origin_data_type_name.scale = numeric_traits_arg.value

        data_format = instance._get_property('dataFormat')
        origin_data_type_name = utils.cdm_data_format_to_syms_data_type(data_format, origin_data_type_name)
        if origin_data_type_name == None:
            return None
        if origin_data_type_name.type_name == None:
            logger.error(ctx, _TAG, 'toData', instance.at_corpus_path, CdmLogCode.ERR_PERSIST_SYMS_UNKNOWN_DATA_FORMAT,
                             instance.display_name)
            return None

        data_col = DataColumn(
        origin_data_type_name = origin_data_type_name,
        name = instance.name)
        return data_col

    @staticmethod
    def create_properties(instance: CdmTypeAttributeDefinition, res_opt: ResolveOptions, options: CopyOptions):
        properties = {}
    
        display_name = instance._get_property('display_name')
        source_name = instance._get_property('source_name')
        description = instance._get_property('description')
        is_read_only = instance._get_property('isReadOnly')
        maximum_length = instance._get_property('maximum_length')
        maximum_value = instance._get_property('maximum_value')
        minimum_value = instance._get_property('minimum_value')
        source_ordering = instance._get_property('source_ordering')
        value_constrained_to_list = instance._get_property('value_constrained_to_list')
        is_primary_key = instance._get_property('is_primary_key')
        def_value = instance._get_property('defaultValue')

        if display_name is not None:
            properties['cdm:display_name'] = display_name

        if instance.explanation is not None:
            properties['cdm:explanation'] = instance.explanation

        if source_name is not None:
            properties['cdm:source_name'] = source_name

        if description is not None:
            properties['cdm:description'] = description

        if instance.applied_traits is not None and len(instance.applied_traits) > 0:
            applied_traits = \
                [trait for trait in instance.applied_traits
                 if isinstance(trait, CdmTraitGroupReference) or not trait.is_from_property] \
                    if instance.applied_traits else None
            properties['cdm:traits'] = copy_data_utils._array_copy_data(res_opt, applied_traits, options)
                       
        if is_read_only is not None:
            properties['cdm:isReadOnly'] = is_read_only
        
        if maximum_length is not None:
            properties['cdm:maximumLength'] = maximum_length
        
        if maximum_value is not None:
            properties['cdm:maximumValue'] = maximum_value
        
        if minimum_value is not None:
            properties['cdm:minimumValue'] = minimum_value

        if source_ordering is not None and source_ordering != 0:
            properties['cdm:sourceOrdering'] = source_ordering

        if value_constrained_to_list is not None:
            properties['cdm:valueConstrainedToList'] = value_constrained_to_list

        if is_primary_key is not None:
            properties['cdm:isPrimaryKey'] = is_primary_key

        if def_value is not None:
            properties['cdm:defaultValue'] = def_value

        return properties